app/core/services/attack_service.py

```
@@ -207,6 +207,218 @@ async def duplicate_attack_service(attack_id: int, db: AsyncSession) -> AttackOu
+
+    Returns:
+        AttackOut: The updated attack.
+
+    Raises:
+        AttackNotFoundError: If attack not found.
+        InvalidStateTransitionError: If transition is not valid.
+    """
+    result = await db.execute(select(Attack).where(Attack.id == attack_id))
+    attack = result.scalar_one_or_none()
+    if not attack:
+        raise AttackNotFoundError(f"Attack {attack_id} not found")
```

Ruff (0.15.0)

[warning] Avoid specifying long messages outside the exception class

(TRY003)